File: backend/app/schemas/course_knowledge.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.enums import FileParseStatus, KnowledgeDocumentStatus
from app.schemas.task import TaskCreateResponse


class CourseFileRead(BaseModel):
    id: int
    original_filename: str
    content_type: str | None = None
    file_size: int
    asset_type: str
    parse_status: FileParseStatus
    created_at: datetime
    updated_at: datetime
    usable_for_course_knowledge: bool = True


class CourseFileListResponse(BaseModel):
    items: list[CourseFileRead]
    total: int


class CourseKnowledgeDocumentCreate(BaseModel):
    file_id: int = Field(gt=0)
    title: str | None = Field(default=None, max_length=255)
    description: str | None = Field(default=None, max_length=1000)


class CourseKnowledgeDocumentRead(BaseModel):
    id: int
    title: str
    file_id: int | None = None
    filename: str | None = None
    status: KnowledgeDocumentStatus
    chunk_count: int
    created_at: datetime
    updated_at: datetime
    owner_name: str | None = None
    course_id: int
    ingest_task_id: int | None = None
    added_to_personal: bool = False
    personal_document_id: int | None = None
    personal_document_status: KnowledgeDocumentStatus | None = None


class CourseKnowledgeDocumentListResponse(BaseModel):
    items: list[CourseKnowledgeDocumentRead]
    total: int
    page: int
    page_size: int


class CourseKnowledgeIngestResponse(BaseModel):
    document_id: int
    status: KnowledgeDocumentStatus
    chunk_count: int
    chroma_collection: str


class CourseKnowledgeDeleteResponse(BaseModel):
    document_id: int
    deleted: bool = True
    deleted_chunks: int = 0


class CourseKnowledgeCopyResponse(BaseModel):
    source_document_id: int
    personal_document_id: int
    personal_file_id: int | None = None
    status: KnowledgeDocumentStatus
    chunk_count: int = 0
    already_added: bool = False


class CourseKnowledgeRetrieveRequest(BaseModel):
    query: str = Field(min_length=1, max_length=2000)
    top_k: int = Field(default=5, ge=1, le=20)
    document_ids: list[int] | None = None


class CourseKnowledgeRetrieveResult(BaseModel):
    content: str
    metadata: dict[str, Any]
    score: float | None = None
    document_id: int | None = None
    title: str | None = None
    filename: str | None = None


class CourseKnowledgeRetrieveResponse(BaseModel):
    query: str
    results: list[CourseKnowledgeRetrieveResult]


class CourseKnowledgeAsyncIngestResponse(TaskCreateResponse):
    pass

File: frontend/src/types/courseKnowledge.ts
export type CourseKnowledgeStatus = 'pending' | 'parsing' | 'ingested' | 'failed' | 'deleted' | string
export type CourseFileParseStatus = 'pending' | 'parsed' | 'failed' | 'deleted' | string

export interface CourseFile {
  id: number
  original_filename: string
  content_type?: string | null
  file_size: number
  asset_type: string
  parse_status: CourseFileParseStatus
  created_at: string
  updated_at: string
  usable_for_course_knowledge: boolean
}

export interface CourseFileListResponse {
  items: CourseFile[]
  total: number
}

export interface CourseKnowledgeDocument {
  id: number
  title: string
  file_id?: number | null
  filename?: string | null
  status: CourseKnowledgeStatus
  chunk_count: number
  created_at: string
  updated_at: string
  owner_name?: string | null
  course_id: number
  ingest_task_id?: number | null
  added_to_personal: boolean
  personal_document_id?: number | null
  personal_document_status?: CourseKnowledgeStatus | null
}

export interface CourseKnowledgeDocumentListResponse {
  items: CourseKnowledgeDocument[]
  total: number
  page: number
  page_size: number
}

export interface CourseKnowledgeDocumentCreatePayload {
  file_id: number
  title?: string | null
  description?: string | null
}

export interface CourseKnowledgeIngestResponse {
  document_id: number
  status: CourseKnowledgeStatus
  chunk_count: number
  chroma_collection: string
}

export interface CourseKnowledgeDeleteResponse {
  document_id: number
  deleted: boolean
  deleted_chunks: number
}

export interface CourseKnowledgeCopyResponse {
  source_document_id: number
  personal_document_id: number
  personal_file_id?: number | null
  status: CourseKnowledgeStatus
  chunk_count: number
  already_added: boolean
}

export interface CourseKnowledgeRetrieveRequest {
  query: string
  top_k?: number
  document_ids?: number[] | null
}

export interface CourseKnowledgeRetrieveResult {
  content: string
  metadata: Record<string, unknown>
  score?: number | null
  document_id?: number | null
  title?: string | null
  filename?: string | null
}

export interface CourseKnowledgeRetrieveResponse {
  query: string
  results: CourseKnowledgeRetrieveResult[]
}
